flow/layers/backdrop_filter_layer.cc

```diff
@@ -11,9 +11,21 @@ BackdropFilterLayer::BackdropFilterLayer(sk_sp<SkImageFilter> filter)
 
 void BackdropFilterLayer::Preroll(PrerollContext* context,
                                   const SkMatrix& matrix) {
+  size_t count = 0;
+  if (context->damage_context) {
+    count = context->damage_context->layer_entries_count();
+  }
   Layer::AutoPrerollSaveLayerState save =
       Layer::AutoPrerollSaveLayerState::Create(context, true, bool(filter_));
   ContainerLayer::Preroll(context, matrix);
+  if (context->damage_context) {
+    // Adjust display bounds to layers below this layer
+    context->damage_context->ApplyImageFilter(0, count, filter_.get(), matrix,
+                                              paint_bounds());
```

Comment:
Why is this not done up above before recursing to children?

Reply:
Because before recursing to children we don't know the bounds. 

However this is not going to work anyway I'm afraid. I don't think we can partially redraw clipped portion of backdrop filter. I didn't really think this through. The problem is, that at the edges of clip, it samples pixels outside of clip, which means it would sample pixels from its own children painted in previous frame.

So right now it seems that the easiest solution to get correct output would be to check if the final damage area intersect any backdrop filter, and if it does, extend it to cover entire backdrop filter. That way we always redraw entire backdrop filter, but only if any layer that needs to be redrawn intersects it.

Reply:
I think I understand what you are getting at, but if the BDF expands the dirty regions below it to include its source region then I think it is OK. It has to do that in a way that is not clipped, though.

For most layers there is a simple "I require no input from the screen and I affect these pixels".
For BDF it has both "I require input from this region (regardless of clipping) and I then affect these pixels (which is subject to the clip)". The trick is to express that "where you get your data from" concept with a different handling of clipping.

Reply:
I see that BDF suffers from a cancerous growth of bounds. If it needs to produce output for `x,y,w,h` then it needs input from `x-d,y-d,w+2d,h+2d`. But, if we then redraw that new rectangle in order to give it the input it needs and we do that in the original framebuffer then that rendering would require input from `x-2d,y-2d,w+4d,h+4d`, rinse and repeat up to the maximum clipped bounds of the BDF output (+d).

If we render into an intermediate buffer instead and then just copy the clipped output into the framebuffer, then we can get away with just one expansion of the bounds. The original need to render `x,y,w,h` would be performed into a surface that was `w+2d,h+2d` in size. When it was done we'd copy out the middle `w,h` part of it into the framebuffer. That requires an extra surface to do, but if the repaint region was tiny then it wouldn't matter.

Alternately, for a clipped BDF, we just repaint its clipped size +d and if that is small enough then doing it in place might be the best choice rather than using an intermediate buffer.

Reply:
Indeed. Rendering everything in w+2d,h+2d surface and then just blitting w,h should fix this. But I was also thinking about another optimization. With damage context we should be able determine, that between past and current frame no layer *below* BDF (that intersects it) needs to be repainted. If that's the case for n subsequent frames, perhaps we could take snapshot of BDF area (after blur and before children) and use it until invalidated. To go even further, if only small are underneath changes, blur it (+d) and and for the rest use the snapshot.

I have to think this through. I think for the next iteration I'll focus on making things render correctly (at the expense of repainting entire BDF if needed).